app/controllers/session_controller.rb

```diff
@@ -271,6 +284,22 @@ def current
     end
   end
 
+  def second_factor
+    RateLimiter.new(nil, "login-hr-#{request.remote_ip}", SiteSetting.max_logins_per_ip_per_hour, 1.hour).performed!
+    RateLimiter.new(nil, "login-min-#{request.remote_ip}", SiteSetting.max_logins_per_ip_per_minute, 1.minute).performed!
+    if user = User.find_by_username_or_email(params[:login])
+      unless user.confirm_password?(params[:password])
+        invalid_credentials
+        return
+      end
+      qrcode = RQRCode::QRCode.new(user.provisioning_uri)
+      qrcode_svg = qrcode.as_svg(offset: 0, color: '000',
```

Comment:
Formatting is a bit weird here. This line has two keys and the others have one. One style I enjoy is like this:

```ruby
some_method(
  offset: 0,
  color: '000'
)
```

That way if the method name changes or you assign it to a variable you don't need to change the indentation of all following lines which pollutes the git history.